### ISSUE TITLE
Issue reported by internet.nl: Insecure DH key exchange parameters

### ISSUE THREAD

**Issue by andrejpodzimek** [svarshavchik/courier#24]
There doesn't seem to be a way to fix this problem:

![image](https://user-images.githubusercontent.com/947737/75844437-4c237c80-5dd6-11ea-9fc9-f372ca582900.png)

I've tried to set `TLS_MIN_DH_BITS=1024` (and `2048` and `4096`), but it seems to have no effect. I couldn't find any parameters that could influence the elliptic curve algorithms (or other choice of key exchange algorithms) selected during key exchnage.

Links from the screenshot:
* [RFC 7919](https://tools.ietf.org/html/rfc7919)
* [IT Security Guidelines for Transport Layer Security (TLS)](https://english.ncsc.nl/publications/publications/2019/juni/01/it-security-guidelines-for-transport-layer-security-tls)
* [ffdhe4096](https://raw.githubusercontent.com/internetstandards/dhe_groups/master/ffdhe4096.pem)
* [ffdhe3072](https://raw.githubusercontent.com/internetstandards/dhe_groups/master/ffdhe3072.pem)
* [ffdhe2048](https://raw.githubusercontent.com/internetstandards/dhe_groups/master/ffdhe2048.pem)

**Comment by svarshavchik** [comment]
As documented, TLS_MIN_DH_BITS is for the GnuTLS library only. If you compiled Courier against GnuTLS, instead of OpenSSL, and see this, then you need to note that. With OpenSSL and GnuTLS, custom DH parameters can be loaded via TLS_DHPARAMS.

**Comment by andrejpodzimek** [comment]
I've figured this out by chance after a long time. Basically the list of ciphers must not contain anything that starts with `DHE-`. In my case, based on [mozilla's security recommendations](https://wiki.mozilla.org/Security/Server_Side_TLS), two ciphers from the recommended *intermediate* list were the culprit: `DHE-RSA-AES128-GCM-SHA256` and `DHE-RSA-AES256-GCM-SHA384` Their removal made this warning go away.

OpenSSL's tricky wildcards are hard to compare against a verbatim list of ciphers; this needs to be checked using `openssl ciphers -v "$TLS_CIPHER_LIST"`.

While this doesn't clarify the root cause, i.e., why the DH params in OpenSSL are configured incorrectly and how they should be configured, it is at least a "workaround" that makes this test pass. (With 4096 bits the problem is still there and 8192 bits take hours to generate, so I suppose that he key size is not the issue here.)

As for GnuTLS, I'm sorry to say that, but each time I tried to build Courier `--with-gnutls`, it either didn't build at all (which was still a lucky case) or crashed (which was still OK and easy to detect) or started hanging at random during authentication (which was not so great). All of this happened years ago and things certainly work better now, but I've wasted so much time on GnuTLS problems that I'm sort of *done with it for the decade*. :grin:
